File: Golang/linkedlist.go
package main

import "fmt"

type ListNode[T comparable] struct {
	value T
	next  *ListNode[T]
}

type LinkedList[T comparable] struct {
	head *ListNode[T]
	tail *ListNode[T]
}

func NewLinkedList[T comparable]() *LinkedList[T] {
	return &LinkedList[T]{}
}

func (l *LinkedList[T]) LPUSHBack(value T) {
	newNode := &ListNode[T]{value: value}
	if l.head == nil {
		l.head = newNode
		l.tail = newNode
		return
	}
	l.tail.next = newNode
	l.tail = newNode
}

func (l *LinkedList[T]) LPUSHFront(value T) {
	newNode := &ListNode[T]{value: value}
	if l.head == nil {
		l.head = newNode
		l.tail = newNode
		return
	}
	newNode.next = l.head
	l.head = newNode
}

func (l *LinkedList[T]) LINSERTAfter(value T, index int) {
	if l.head == nil || index < 1 {
		return
	}

	iter := l.head
	for i := 1; i < index; i++ {
		if iter == nil {
			return
		}
		iter = iter.next
	}

	if iter == nil {
		return
	}

	newNode := &ListNode[T]{value: value}
	newNode.next = iter.next
	iter.next = newNode

	if iter == l.tail {
		l.tail = newNode
	}
}

func (l *LinkedList[T]) LINSERTBefore(value T, index int) {
	if l.head == nil || index < 1 {
		return
	}

	if index == 1 {
		l.LPUSHFront(value)
		return
	}

	iter := l.head
	for i := 1; i < index-1; i++ {
		if iter == nil {
			return
		}
		iter = iter.next
	}

	if iter == nil {
		return
	}

	newNode := &ListNode[T]{value: value}
	newNode.next = iter.next
	iter.next = newNode
}

func (l *LinkedList[T]) LDELHead() {
	if l.head == nil {
		return
	}
	l.head = l.head.next
	if l.head == nil {
		l.tail = nil
	}
}

func (l *LinkedList[T]) LDELTail() {
	if l.head == nil {
		return
	}

	if l.head == l.tail {
		l.head = nil
		l.tail = nil
		return
	}

	iter := l.head
	for iter.next != l.tail {
		iter = iter.next
	}

	iter.next = nil
	l.tail = iter
}

func (l *LinkedList[T]) LDELAfter(index int) {
	if l.head == nil || index < 1 {
		return
	}

	iter := l.head
	for i := 1; i < index; i++ {
		if iter == nil {
			return
		}
		iter = iter.next
	}

	if iter == nil || iter.next == nil {
		return
	}

	iter.next = iter.next.next
	if iter.next == nil {
		l.tail = iter
	}
}

func (l *LinkedList[T]) LDELBefore(index int) {
	if l.head == nil || index <= 1 {
		return
	}

	if index == 2 {
		l.LDELHead()
		return
	}

	iter := l.head
	for i := 1; i < index-2; i++ {
		if iter == nil {
			return
		}
		iter = iter.next
	}

	if iter == nil || iter.next == nil {
		return
	}

	iter.next = iter.next.next
}

func (l *LinkedList[T]) LDEL(value T) {
	if l.head == nil {
		return
	}

	if l.head.value == value {
		l.LDELHead()
		return
	}

	iter := l.head
	for iter.next != nil && iter.next.value != value {
		iter = iter.next
	}

	if iter.next == nil {
		return
	}

	iter.next = iter.next.next
	if iter.next == nil {
		l.tail = iter
	}
}

func (l *LinkedList[T]) Print() {
	current := l.head
	for current != nil {
		fmt.Print(current.value, " ")
		current = current.next
	}
	fmt.Println()
}

func (l *LinkedList[T]) LGET(value T) int {
	index := 1
	current := l.head

	for current != nil {
		if current.value == value {
			return index
		}
		current = current.next
		index++
	}
	return -1
}

File: C++/LinkedList.cpp
#include <iostream>

#include "LinkedList.h"

using namespace std;

void LinkedList::LPUSHBack(const string &value) {
    auto* newNode = new ListNode(value);
    if (head == nullptr) {
        head = newNode;
        tail = newNode;
    }
    tail->next = newNode;
    tail = newNode;
}

void LinkedList::LPUSHFront(const string& value) {
    auto* newNode = new ListNode(value);
    if (head == nullptr) {
        head = newNode;
        tail = newNode;
    }
    newNode->next = head;
    head = newNode;
}

void LinkedList::LINSERTAfter(const string& value, const int index) {
    auto* newNode = new ListNode(value);
    auto* iter = head;
    for (int i = 1; i < index; i++ ) {
        iter = iter->next;
    }
    if (!iter) return;
    newNode->next = iter->next;
    iter->next = newNode;
}

void LinkedList::LINSERTBefore(const string& value, const int index) {
    if (head == nullptr) return;

    if (index == 1) LPUSHFront(value);

    auto* iter = head;
    for (int i = 1; i < index - 1; i++ ) {
        iter = iter->next;
    }

    if (!iter) return;

    auto* newNode = new ListNode(value);

    newNode->next = iter->next;
    iter->next = newNode;
}

void LinkedList::LDELHead() {
    if (head == nullptr) return;

    auto* nodeToDelete = head;
    head = head->next;
    delete nodeToDelete;
}


void LinkedList::LDELTail() {
    if (head == nullptr) return;

    if (head == tail) {
        delete head;
        head = tail = nullptr;
        return;
    }

    auto* iter = head;
    while (iter->next != tail) {
        iter = iter->next;
    }

    auto* nodeToDelete = tail;
    tail = iter;
    tail->next = nullptr;
    delete nodeToDelete;
}

void LinkedList::LDELAfter(const int index) {
    if (head == nullptr) return;

    auto* iter = head;
    for (int i = 1; i < index; i++ ) {
        iter = iter->next;
    }
    if (!iter) return;

    auto* nodeToDelete = iter->next;
    iter->next = nodeToDelete->next;

    delete nodeToDelete;
}

void LinkedList::LDELBefore(const int index) {
    if (head == nullptr or index <= 1) return;

    if (index == 2) {
        auto* nodeToDelete = head;
        head = head->next;
        delete nodeToDelete;
        return;
    }

    auto* iter = head;

    for (int i = 1; i < index - 2; i++ ) {
        iter = iter->next;
    }

    if (!iter) return;

    auto* nodeToDelete = iter->next;
    iter->next = nodeToDelete->next;

    delete nodeToDelete;
}

void LinkedList::LDEL(const string& value) {
    if (head == nullptr) return;

    if (head->value == value) LDELHead();
    if (tail->value == value) LDELTail();

    auto* iter = head;

    while (iter->next->value != value) {
        iter = iter->next;
    }

    auto* nodeToDelete = iter;
    iter->next = nodeToDelete->next;

    delete nodeToDelete;
}

void LinkedList::print() {
    auto* current = head;
    while (current != nullptr) {
        cout << current->value << " ";
        current = current->next;
    }
    cout << endl;
}

int LinkedList::LGET(const string& value) {
    int index = 1;
    auto* current = head;

    while (current->next!= nullptr) {
        if (current->value == value) {
            return index;
        }
        current = current->next;
        index++;
    }

    return -1;
}